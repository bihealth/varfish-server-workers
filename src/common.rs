//! Common functionality.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    ops::Range,
    path::Path,
};

use byte_unit::Byte;
use clap_verbosity_flag::{InfoLevel, Verbosity};

use clap::Parser;
use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};
use hgvs::static_data::Assembly;
use indexmap::IndexMap;

/// Commonly used command line arguments.
#[derive(Parser, Debug)]
pub struct Args {
    /// Verbosity of the program
    #[clap(flatten)]
    pub verbose: Verbosity<InfoLevel>,
}

/// Helper to print the current memory resident set size via `tracing`.
pub fn trace_rss_now() {
    let me = procfs::process::Process::myself().unwrap();
    let page_size = procfs::page_size();
    tracing::debug!(
        "RSS now: {}",
        Byte::from_bytes((me.stat().unwrap().rss * page_size) as u128).get_appropriate_unit(true)
    );
}

/// Definition of canonical chromosome names.
pub const CHROMS: &[&str] = &[
    "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17",
    "18", "19", "20", "21", "22", "X", "Y", "M",
];

/// Build mapping of chromosome names to chromosome counts.
pub fn build_chrom_map() -> IndexMap<String, usize> {
    let mut result = IndexMap::new();
    for (i, &chrom_name) in CHROMS.iter().enumerate() {
        result.insert(chrom_name.to_owned(), i);
        result.insert(format!("chr{chrom_name}").to_owned(), i);
    }
    result.insert("x".to_owned(), 22);
    result.insert("y".to_owned(), 23);
    result.insert("chrx".to_owned(), 22);
    result.insert("chry".to_owned(), 23);
    result.insert("mt".to_owned(), 24);
    result.insert("m".to_owned(), 24);
    result.insert("chrmt".to_owned(), 24);
    result.insert("chrm".to_owned(), 24);
    result.insert("MT".to_owned(), 24);
    result.insert("chrMT".to_owned(), 24);
    result
}

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently opena  file with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// Compute reciprocal overlap between two ranges.
pub fn reciprocal_overlap(lhs: Range<i32>, rhs: Range<i32>) -> f32 {
    let lhs_b = lhs.start;
    let lhs_e = lhs.end;
    let rhs_b = rhs.start;
    let rhs_e = rhs.end;
    let ovl_b = std::cmp::max(lhs_b, rhs_b);
    let ovl_e = std::cmp::min(lhs_e, rhs_e);
    if ovl_b >= ovl_e {
        0f32
    } else {
        let ovl_len = (ovl_e - ovl_b) as f32;
        let x1 = ovl_len / (lhs_e - lhs_b) as f32;
        let x2 = ovl_len / (rhs_e - rhs_b) as f32;
        x1.min(x2)
    }
}

/// Helper to convert ENSEMBL and RefSeq gene ID to u32.
pub fn numeric_gene_id(raw_id: &str) -> Result<u32, anyhow::Error> {
    let clean_id = if raw_id.starts_with("ENSG") {
        // Strip "ENSG" prefix and as many zeroes as follow
        raw_id
            .chars()
            .skip("ENSG".len())
            .skip_while(|c| *c == '0')
            .collect()
    } else {
        raw_id.to_owned()
    };

    clean_id
        .parse::<u32>()
        .map_err(|e| anyhow::anyhow!("could not parse gene id {:?}: {}", &clean_id, &e))
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

/// Select the genome release to use.
#[derive(
    clap::ValueEnum,
    Clone,
    Copy,
    Debug,
    strum::Display,
    PartialEq,
    Eq,
    enum_map::Enum,
    PartialOrd,
    Ord,
    Hash,
)]
pub enum GenomeRelease {
    // GRCh37 / hg19
    #[strum(serialize = "GRCh37")]
    Grch37,
    /// GRCh38 / hg38
    #[strum(serialize = "GRCh38")]
    Grch38,
}

impl GenomeRelease {
    pub fn name(&self) -> String {
        match self {
            GenomeRelease::Grch37 => String::from("GRCh37"),
            GenomeRelease::Grch38 => String::from("GRCh38"),
        }
    }
}

impl From<GenomeRelease> for Assembly {
    fn from(val: GenomeRelease) -> Self {
        match val {
            GenomeRelease::Grch37 => Assembly::Grch37p10,
            GenomeRelease::Grch38 => Assembly::Grch38,
        }
    }
}

impl From<Assembly> for GenomeRelease {
    fn from(assembly: Assembly) -> Self {
        match assembly {
            Assembly::Grch37 | Assembly::Grch37p10 => GenomeRelease::Grch37,
            Assembly::Grch38 => GenomeRelease::Grch38,
        }
    }
}

impl std::str::FromStr for GenomeRelease {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_ascii_lowercase();
        if s.starts_with("grch37") {
            Ok(GenomeRelease::Grch37)
        } else if s.starts_with("grch38") {
            Ok(GenomeRelease::Grch38)
        } else {
            Err(anyhow::anyhow!("Unknown genome release: {}", s))
        }
    }
}

/// The version of `varfish-server-worker` package.
pub const VERSION: &str = env!("CARGO_PKG_VERSION");

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn numeric_gene_id_simple() -> Result<(), anyhow::Error> {
        assert_eq!(1, numeric_gene_id("ENSG0000000001")?);
        assert_eq!(1, numeric_gene_id("ENSG1")?);
        assert_eq!(1, numeric_gene_id("1")?);

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[test]
    fn trace_rss_now_smoke() {
        super::trace_rss_now();
    }

    #[test]
    fn build_chrom_map_snapshot() {
        let map = super::build_chrom_map();
        insta::assert_yaml_snapshot!(map);
    }

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_read_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        let mut f = super::open_read_maybe_gz(if is_gzip {
            "tests/common/test.txt.gz"
        } else {
            "tests/common/test.txt"
        })?;

        let mut buf = String::new();
        f.read_to_string(&mut buf)?;

        insta::assert_snapshot!(&buf);

        Ok(())
    }

    #[rstest::rstest]
    #[case(0..10, 0..10, 1.0)]
    #[case(0..10, 5..15, 0.5)]
    #[case(5..15, 0..10, 0.5)]
    #[case(0..10, 10..20, 0.0)]
    #[case(0..2, 0..10, 0.2)]
    #[case(0..10, 0..2, 0.2)]
    fn reciprocal_overlap(
        #[case] lhs: std::ops::Range<i32>,
        #[case] rhs: std::ops::Range<i32>,
        #[case] expected: f32,
    ) {
        let actual = super::reciprocal_overlap(lhs, rhs);
        assert!(float_cmp::approx_eq!(f32, expected, actual, ulps = 2))
    }

    #[rstest::rstest]
    #[case("ENSG0000000142", 142)]
    #[case("42", 42)]
    fn numeric_gene_id(#[case] raw_id: &str, #[case] expected: u32) {
        let actual = super::numeric_gene_id(raw_id).unwrap();
        assert_eq!(expected, actual);
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "GRCh37")]
    #[case(crate::common::GenomeRelease::Grch38, "GRCh38")]
    fn genome_release_name(#[case] release: super::GenomeRelease, #[case] expected: &str) {
        assert_eq!(expected, release.name());
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        hgvs::static_data::Assembly::Grch38
    )]
    fn assembly_from_genome_release(
        #[case] release: super::GenomeRelease,
        #[case] assembly: hgvs::static_data::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: hgvs::static_data::Assembly = release.into();

        assert_eq!(res, assembly);

        Ok(())
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37
    )]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        hgvs::static_data::Assembly::Grch38
    )]
    fn genome_release_from_assembly(
        #[case] release: super::GenomeRelease,
        #[case] assembly: hgvs::static_data::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = assembly.into();

        assert_eq!(res, release);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "grch37")]
    #[case(crate::common::GenomeRelease::Grch38, "grch38")]
    fn genome_relese_from_str(
        #[case] release: super::GenomeRelease,
        #[case] s: &str,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = s.parse()?;

        assert_eq!(res, release);

        Ok(())
    }
}
