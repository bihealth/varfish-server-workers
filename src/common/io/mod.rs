//! Common, IO-related code.

pub mod std;
