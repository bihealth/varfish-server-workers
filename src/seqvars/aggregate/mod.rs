//! Implementation of `seqvars aggregate` subcommand.

pub mod ds;

use mehari::common::noodles::open_vcf_reader;
use noodles::vcf;
use rayon::prelude::*;
use std::{str::FromStr as _, sync::Arc};

use crate::common::{self, genotype_to_string, Chrom, Genotype};

/// Command line arguments for `seqvars aggregate` subcommand.
#[derive(Debug, clap::Parser)]
#[command(author, version, about = "ingest sequence variant VCF", long_about = None)]
pub struct Args {
    /// The assumed genome build.
    #[clap(long)]
    pub genomebuild: crate::common::GenomeRelease,
    /// Path to the output RocksDB.
    #[clap(long)]
    pub path_out_rocksdb: String,
    /// Path to input VCF file(s).
    #[clap(long)]
    pub path_input: Vec<String>,

    /// Column family name for the count data.
    #[clap(long, default_value = "counts")]
    pub cf_counts: String,
    /// Column family name for the carrier UUID data.
    #[clap(long, default_value = "carriers")]
    pub cf_carriers: String,
    /// Set the number of threads to use, defaults to number of cores.
    #[clap(long)]
    pub num_threads: Option<usize>,

    /// Optional path to RocksDB WAL directory.
    #[arg(long)]
    pub path_wal_dir: Option<String>,
}

/// Returns whether the given coordinate is in PAR for `chrom`, `pos` (1-based) and `genombuild`.
fn is_par(chrom: Chrom, pos: usize, genomebuild: crate::common::GenomeRelease) -> bool {
    match (chrom, genomebuild) {
        (Chrom::X, crate::common::GenomeRelease::Grch37) => {
            (60001..=2699520).contains(&pos) || (154931044..=155260560).contains(&pos)
        }
        (Chrom::X, crate::common::GenomeRelease::Grch38) => {
            (10001..=2781479).contains(&pos) || (155701383..=156030895).contains(&pos)
        }
        (Chrom::Y, crate::common::GenomeRelease::Grch37) => {
            (10001..=2649520).contains(&pos) || (59034050..=59363566).contains(&pos)
        }
        (Chrom::Y, crate::common::GenomeRelease::Grch38) => {
            (10001..=2781479).contains(&pos) || (56887903..=57217415).contains(&pos)
        }
        _ => false,
    }
}

/// Extract counts and carrier data from a single VCF record.
fn handle_record(
    input_record: &vcf::variant::RecordBuf,
    input_header: &vcf::Header,
    pedigree: &mehari::ped::PedigreeByName,
    case_uuid: &uuid::Uuid,
    genomebuild: crate::common::GenomeRelease,
) -> Result<(ds::Counts, ds::CarrierList), anyhow::Error> {
    let chrom: Chrom = annonars::common::cli::canonicalize(
        input_record.reference_sequence_name().to_string().as_str(),
    )
    .as_str()
    .parse()?;

    let mut res_counts = ds::Counts::default();
    let mut res_carriers = ds::CarrierList::default();

    let start: usize = input_record
        .variant_start()
        .ok_or_else(|| anyhow::anyhow!("missing variant start in record {:?}", &input_record))?
        .into();

    for (name, sample) in input_header
        .sample_names()
        .iter()
        .zip(input_record.samples().values())
    {
        let individual = pedigree
            .individuals
            .get(name)
            .ok_or_else(|| anyhow::anyhow!("individual {} not found in pedigree", name))?;

        use noodles::vcf::variant::record::samples::keys::key;
        let genotype = if let Some(Some(
            vcf::variant::record_buf::samples::sample::value::Value::Genotype(gt),
        )) = sample.get(key::GENOTYPE)
        {
            Genotype::from_str(&genotype_to_string(&gt)?)?
        } else {
            anyhow::bail!("invalid genotype value in {:?}", &sample)
        };

        // Ac-hoc enum for readable PAR status.
        #[derive(Debug, PartialEq, Eq)]
        enum _IsPar {
            IsPar,
            NoPar,
        }
        use _IsPar::*;
        let is_par = if is_par(chrom, start, genomebuild) {
            IsPar
        } else {
            NoPar
        };

        let carrier_genotype = match (chrom, is_par, individual.sex, genotype) {
            (_, _, _, Genotype::WithNoCall) => continue,
            // On the autosomes, male/female are handled the same.
            (Chrom::Auto, _, _, Genotype::HomRef) => {
                res_counts.count_homref += 1;
                ds::Genotype::HomRef
            }
            (Chrom::Auto, _, _, Genotype::Het) => {
                res_counts.count_het += 1;
                ds::Genotype::Het
            }
            (Chrom::Auto, _, _, Genotype::HomAlt) => {
                res_counts.count_homalt += 2;
                ds::Genotype::HomAlt
            }
            // On the gonomosomes, we handle call male variant calls as hemizygous outside PAR.
            (Chrom::X, NoPar, mehari::ped::Sex::Male, Genotype::HomRef)
            | (Chrom::Y, NoPar, mehari::ped::Sex::Male, Genotype::HomRef) => {
                res_counts.count_hemiref += 1;
                ds::Genotype::HemiRef
            }
            (Chrom::X, NoPar, mehari::ped::Sex::Male, Genotype::Het)
            | (Chrom::Y, NoPar, mehari::ped::Sex::Male, Genotype::Het)
            | (Chrom::X, NoPar, mehari::ped::Sex::Male, Genotype::HomAlt)
            | (Chrom::Y, NoPar, mehari::ped::Sex::Male, Genotype::HomAlt) => {
                res_counts.count_hemialt += 1;
                ds::Genotype::HemiAlt
            }
            // For female samples, we handle chrX as biallelic, same as male inside PAR on
            // chrX/chrY.  Note that read mapping pipelines N-mask the PAR on chrY, so we
            // don't count twice.
            (Chrom::X, IsPar, mehari::ped::Sex::Male, Genotype::HomRef)
            | (Chrom::Y, IsPar, mehari::ped::Sex::Male, Genotype::HomRef)
            | (Chrom::X, _, mehari::ped::Sex::Female, Genotype::HomRef) => {
                res_counts.count_homref += 1;
                ds::Genotype::HomRef
            }
            (Chrom::X, IsPar, mehari::ped::Sex::Male, Genotype::Het)
            | (Chrom::Y, IsPar, mehari::ped::Sex::Male, Genotype::Het)
            | (Chrom::X, _, mehari::ped::Sex::Female, Genotype::Het) => {
                res_counts.count_het += 1;
                ds::Genotype::Het
            }
            (Chrom::X, IsPar, mehari::ped::Sex::Male, Genotype::HomAlt)
            | (Chrom::Y, IsPar, mehari::ped::Sex::Male, Genotype::HomAlt)
            | (Chrom::X, _, mehari::ped::Sex::Female, Genotype::HomAlt) => {
                res_counts.count_homalt += 2;
                ds::Genotype::HomAlt
            }
            // We ignore calls to chrY for female samples.
            (Chrom::Y, _, mehari::ped::Sex::Female, Genotype::HomRef)
            | (Chrom::Y, _, mehari::ped::Sex::Female, Genotype::Het)
            | (Chrom::Y, _, mehari::ped::Sex::Female, Genotype::HomAlt) => ds::Genotype::HomRef,
            // Do not count samples with unknown sex on gonomosomes.
            (Chrom::X, _, mehari::ped::Sex::Unknown, _)
            | (Chrom::Y, _, mehari::ped::Sex::Unknown, _) => ds::Genotype::HomRef,
        };

        if carrier_genotype != ds::Genotype::HomRef {
            res_carriers.carriers.push(ds::Carrier {
                uuid: *case_uuid,
                index: pedigree
                    .individuals
                    .get_index_of(name.as_str())
                    .ok_or_else(|| anyhow::anyhow!("individual {} not found in pedigree", &name))?
                    as u8,
                genotype: carrier_genotype,
            });
        }
    }

    Ok((res_counts, res_carriers))
}

/// Import one VCF file into the database.
///
/// This function is `async` because we potentially need to read from S3.
async fn import_vcf(
    db: &Arc<rocksdb::TransactionDB<rocksdb::MultiThreaded>>,
    path_input: &str,
    cf_counts: &str,
    cf_carriers: &str,
    genomebuild: crate::common::GenomeRelease,
) -> Result<(), anyhow::Error> {
    let mut input_reader = open_vcf_reader(path_input)
        .await
        .map_err(|e| anyhow::anyhow!("could not open file {} for reading: {}", path_input, e))?;
    let input_header = input_reader.read_header().await?;

    let cf_counts = db.cf_handle(cf_counts).expect("checked earlier");
    let cf_carriers = db.cf_handle(cf_carriers).expect("checked earlier");

    let (pedigree, case_uuid) = common::extract_pedigree_and_case_uuid(&input_header)?;
    let mut prev = std::time::Instant::now();
    let mut record_buf = vcf::variant::RecordBuf::default();
    loop {
        let bytes_read = input_reader
            .read_record_buf(&input_header, &mut record_buf)
            .await
            .map_err(|e| anyhow::anyhow!("problem reading VCF file {}: {}", path_input, e))?;
        if bytes_read == 0 {
            break; // EOF
        }

        // Obtain counts from the current variant.
        let (this_counts_data, this_carrier_data) = handle_record(
            &record_buf,
            &input_header,
            &pedigree,
            &case_uuid,
            genomebuild,
        )?;
        // Obtain annonars variant key from current allele for RocksDB lookup.
        let vcf_var = annonars::common::keys::Var::from_vcf_allele(&record_buf, 0);
        let key: Vec<u8> = vcf_var.clone().into();

        let max_retries = 10;
        let mut retries = 0;
        while retries < max_retries {
            let this_counts_data = this_counts_data.clone();
            let this_carrier_data = this_carrier_data.clone();

            let transaction = db.transaction();

            // Read data for variant from database.
            let mut db_counts_data = transaction.get_cf(&cf_counts, key.clone()).map_err(|e| {
                    anyhow::anyhow!(
                        "problem acessing counts data for variant {:?}: {} (non-existing would be fine)",
                        &vcf_var,
                        e
                    )
                })?.map(|buffer| ds::Counts::from_vec(&buffer)).unwrap_or_default();
            let mut db_carrier_data = transaction.get_cf(&cf_carriers, key.clone()).map_err(|e| {
                    anyhow::anyhow!(
                        "problem acessing carrier data for variant {:?}: {} (non-existing would be fine)",
                        &vcf_var,
                        e
                    )
                })?
                .map(|buffer| ds::CarrierList::try_from(buffer.as_slice()))
                .transpose()
                .map_err(|e| {
                    anyhow::anyhow!(
                        "problem decoding carrier data for variant {:?}: {}",
                        &vcf_var,
                        e
                    )
                })?
                .unwrap_or_default();

            // Aggregate the data.
            db_counts_data.aggregate(this_counts_data);
            db_carrier_data.aggregate(this_carrier_data);

            // Write data for variant back to database.
            transaction
                .put_cf(&cf_counts, key.clone(), &db_counts_data.to_vec())
                .map_err(|e| {
                    anyhow::anyhow!(
                        "problem writing counts data for variant {:?}: {}",
                        &vcf_var,
                        e
                    )
                })?;
            transaction
                .put_cf(&cf_carriers, key.clone(), &db_carrier_data.to_vec())
                .map_err(|e| {
                    anyhow::anyhow!(
                        "problem writing carrier data for variant {:?}: {}",
                        &vcf_var,
                        e
                    )
                })?;

            let res = transaction.commit();
            match res {
                Ok(_) => break,
                Err(e) => {
                    retries += 1;
                    if retries > 5 {
                        tracing::warn!(
                            "problem committing transaction for variant {:?}: {} (retry #{})",
                            &vcf_var,
                            e,
                            retries
                        );
                    }
                }
            }
        }
        if retries >= max_retries {
            return Err(anyhow::anyhow!(
                "problem committing transaction for variant {:?}: {} (max retries exceeded)",
                &vcf_var,
                retries
            ));
        }

        // Write out progress indicator every 60 seconds.
        if prev.elapsed().as_secs() >= 60 {
            tracing::info!("at {:?}", &vcf_var);
            prev = std::time::Instant::now();
        }
    }

    Ok(())
}

/// Perform the parallel import of VCF files.
fn vcf_import(
    db: &Arc<rocksdb::TransactionDB<rocksdb::MultiThreaded>>,
    path_input: &[&str],
    cf_counts: &str,
    cf_carriers: &str,
    genomebuild: crate::common::GenomeRelease,
) -> Result<(), anyhow::Error> {
    path_input
        .par_iter()
        .map(|path_input| {
            // We create a Tokio scheduler for the current file as we need it
            // to wait / block for the VCF import running in the current Rayon
            // thread.
            tokio::runtime::Builder::new_current_thread()
                .build()
                .map_err(|e| {
                    anyhow::anyhow!(
                        "building Tokio runtime for VCF file {} failed: {}",
                        path_input,
                        e
                    )
                })?
                .block_on(import_vcf(
                    db,
                    path_input,
                    cf_counts,
                    cf_carriers,
                    genomebuild,
                ))
                .map_err(|e| anyhow::anyhow!("processing VCF file {} failed: {}", path_input, e))
        })
        .collect::<Result<Vec<_>, _>>()
        .map(|_| ())
}

/// Main entry point for `seqvars aggregate` sub command.
pub fn run(args_common: &crate::common::Args, args: &Args) -> Result<(), anyhow::Error> {
    let before_anything = std::time::Instant::now();
    tracing::info!("args_common = {:#?}", &args_common);
    tracing::info!("args = {:#?}", &args);

    if let Some(num_threads) = args.num_threads {
        rayon::ThreadPoolBuilder::new()
            .num_threads(num_threads)
            .build_global()
            .map_err(|e| anyhow::anyhow!("building global Rayon thread pool failed: {}", e))?;
    }

    common::trace_rss_now();

    // Build path of all input files to read, read through files given by `@path`.
    let path_input = args
        .path_input
        .iter()
        .flat_map(|path| {
            if path.starts_with('@') {
                std::fs::read_to_string(path.trim_start_matches('@'))
                    .expect("checked above")
                    .lines()
                    .map(|line| line.trim())
                    .filter(|line| !line.is_empty())
                    .map(|line| line.to_string())
                    .collect::<Vec<_>>()
            } else {
                vec![path.clone()]
            }
        })
        .collect::<Vec<_>>();

    tracing::info!("Opening RocksDB...");
    let options = rocksdb_utils_lookup::tune_options(
        rocksdb::Options::default(),
        args.path_wal_dir.as_ref().map(|s| s.as_ref()),
    );
    let tx_options = rocksdb::TransactionDBOptions::default();
    let cf_names = &["meta", &args.cf_counts, &args.cf_carriers];
    let cf_descriptors = cf_names
        .iter()
        .map(|name| rocksdb::ColumnFamilyDescriptor::new(*name, options.clone()))
        .collect::<Vec<_>>();

    // scope for the transaction database
    {
        let db: Arc<rocksdb::TransactionDB<rocksdb::MultiThreaded>> =
            Arc::new(rocksdb::TransactionDB::open_cf_descriptors(
                &options,
                &tx_options,
                &args.path_out_rocksdb,
                cf_descriptors,
            )?);
        tracing::info!("  writing meta information");
        let cf_meta = db.cf_handle("meta").unwrap();
        db.put_cf(&cf_meta, "varfish-worker-version", common::worker_version())?;
        db.put_cf(&cf_meta, "db-name", "seqvars-aggregation")?;
        tracing::info!("... done opening RocksDB");

        tracing::info!("Importing VCF files ...");
        let before_import = std::time::Instant::now();
        let paths = path_input.iter().map(|s| s.as_ref()).collect::<Vec<_>>();
        vcf_import(
            &db,
            &paths,
            &args.cf_counts,
            &args.cf_carriers,
            args.genomebuild,
        )?;
        tracing::info!(
            "... done importing VCF files in {:?}",
            before_import.elapsed()
        );
    }

    // scope for compaction
    {
        let db = Arc::new(rocksdb::DB::open_cf_with_opts(
            &options,
            &args.path_out_rocksdb,
            cf_names
                .iter()
                .map(|name| (name.to_string(), options.clone()))
                .collect::<Vec<_>>(),
        )?);
        tracing::info!("Running RocksDB compaction ...");
        let before_compaction = std::time::Instant::now();
        rocksdb_utils_lookup::force_compaction_cf(&db, cf_names, Some("  "), true)?;
        tracing::info!(
            "... done compacting RocksDB in {:?}",
            before_compaction.elapsed()
        );
    }

    tracing::info!(
        "All of `seqvars aggregate` completed in {:?}",
        before_anything.elapsed()
    );
    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_is_par() {
        assert_eq!(
            super::is_par(super::Chrom::X, 60000, crate::common::GenomeRelease::Grch37),
            false
        );
        assert_eq!(
            super::is_par(super::Chrom::X, 60001, crate::common::GenomeRelease::Grch37),
            true
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                2699520,
                crate::common::GenomeRelease::Grch37
            ),
            true
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                2699521,
                crate::common::GenomeRelease::Grch37
            ),
            false
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                154931043,
                crate::common::GenomeRelease::Grch37
            ),
            false
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                154931044,
                crate::common::GenomeRelease::Grch37
            ),
            true
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                155260560,
                crate::common::GenomeRelease::Grch37
            ),
            true
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                155260561,
                crate::common::GenomeRelease::Grch37
            ),
            false
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                155260561,
                crate::common::GenomeRelease::Grch38
            ),
            false
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                155701383,
                crate::common::GenomeRelease::Grch38
            ),
            true
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                156030895,
                crate::common::GenomeRelease::Grch38
            ),
            true
        );
        assert_eq!(
            super::is_par(
                super::Chrom::X,
                156030896,
                crate::common::GenomeRelease::Grch38
            ),
            false
        );
        assert_eq!(
            super::is_par(super::Chrom::Y, 10000, crate::common::GenomeRelease::Grch37),
            false
        );
        assert_eq!(
            super::is_par(super::Chrom::Y, 10001, crate::common::GenomeRelease::Grch37),
            true
        );
    }

    #[tracing_test::traced_test]
    #[test]
    fn handle_record_snapshot() -> Result<(), anyhow::Error> {
        let path = "tests/seqvars/aggregate/ingest.vcf";
        let mut vcf_reader = vcf::io::reader::Builder::default()
            .build_from_path(path)
            .unwrap();
        let header = vcf_reader.read_header().unwrap();

        let mut record_buf = vcf::variant::RecordBuf::default();
        loop {
            let bytes_read = vcf_reader
                .read_record_buf(&header, &mut record_buf)
                .map_err(|e| anyhow::anyhow!("problem reading VCF file {}: {}", path, e))?;
            if bytes_read == 0 {
                break; // EOF
            }

            let (pedigree, case_uuid) = common::extract_pedigree_and_case_uuid(&header)?;
            let (counts, carriers) = super::handle_record(
                &record_buf,
                &header,
                &pedigree,
                &case_uuid,
                crate::common::GenomeRelease::Grch37,
            )?;

            insta::assert_debug_snapshot!(counts);
            insta::assert_debug_snapshot!(carriers);
        }

        Ok(())
    }
}
