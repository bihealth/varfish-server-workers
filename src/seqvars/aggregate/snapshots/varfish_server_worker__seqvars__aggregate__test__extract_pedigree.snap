---
source: src/seqvars/aggregate/mod.rs
expression: pedigree
---
PedigreeByName {
    individuals: {},
}
